### PR TITLE
OCPBUGS-33428: Reconcile KAS endpoints and endpoint slice

### PR DESCRIPTION
**Pull request by rtheis**
**What this PR does / why we need it**:

See issue for details and test results.

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes # https://issues.redhat.com/browse/OCPBUGS-33428

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [x] This change includes unit tests.

### CODE REVIEW

**Review by stevekuznetsov — COMMENTED** [review]
Code seems straightforward - is this associated with a Jira ticket?

**Comment by rtheis** [comment]
> Code seems straightforward - is this associated with a Jira ticket?

@stevekuznetsov Thanks for the review.  I will address the comments and will be opening a Jira with the details.  I wanted to pull together some test results with background information as to why I think we need this change.

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/kas/reconcile.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package kas
+
+import (
+	"github.com/openshift/hypershift/support/util"
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	"k8s.io/utils/ptr"
+)
+
+func ReconcileKASEndpoints(endpoints *corev1.Endpoints, address string, port int32) error {
```
[the remark this answers is not in the document]
Comment (rtheis) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/kas/reconcile.go]: I've updated the signatures.

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go [hunk 2]
```diff
@@ -275,6 +278,16 @@ func (r *reconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result
 		}); err != nil {
 			errs = append(errs, fmt.Errorf("failed to reconcile kubernetes.default endpoints: %w", err))
 		}
+	} else {
+		// Clusters with "none" as their Kubernetes API server endpoint reconciliation
```
[the remark this answers is not in the document]
Comment (rtheis) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go]: I've refactored to handle both cases with one reconciliation.

**Comment by rtheis** [comment]
/retitle OCPBUGS-33428: Reconcile KAS endpoints and endpoint slice

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go [hunk 3]
```diff
@@ -1163,6 +1157,39 @@ func (r *reconciler) reconcileOpenshiftOAuthAPIServerAPIServices(ctx context.Con
 	return errors.NewAggregate(errs)
 }
 
+func (r *reconciler) reconcileKASEndpoints(ctx context.Context, hcp *hyperv1.HostedControlPlane) error {
```
Comment (stevekuznetsov) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1160]: I really wish we had some way to test the whole reconciler together so we could capture that it now does this. @sjenning

**Review by stevekuznetsov — APPROVED** [review]
/lgtm

**Comment by enxebre** [comment]
Why is this issue only hit when having haproxy to intercept traffic to the endpoint? why is it not hit with a traditional svc->endpoint->pod flow?
Is it because we set advertise-address explicitly?

**Comment by rtheis** [comment]
> Why is this issue only hit when having haproxy to intercept traffic to the endpoint? why is it not hit with a traditional svc->endpoint->pod flow? Is it because we set advertise-address explicitly and that results in a single address for the endpoint hence the issue in the ticket when one single instance goes down, or is it something else?
> 
> I'm concern we opt out to other goodies by setting none for --endpoint-reconciler-type
> 
> Other than the concerns above, code changes lgtm

@enxebre My understanding is that setting the KAS advertise-address results in a single address for the KAS endpoint in the data plane regardless of there being multiple KAS pods/endpoints in the control plane.  As a result, when any KAS pod is terminated, the pod will now remove its address from the endpoint. In previous releases, the KAS would see a single address and refuse to remove its address from the endpoint.  That failsafe was removed by Kubernetes and thus we see single endpoint setups, such as used by HyperShift, encounter this KAS in cluster availability problem during routine maintenance activities.  Based on our discussions with the Kubernetes community via https://github.com/kubernetes/kubernetes/issues/118777, I think we unfortunately need to set `none` for `--endpoint-reconciler-type` and manage the KAS endpoints and endpoint slice ourselves.

**Pull request merged**